### ISSUE TITLE
How "public static class Pool<T> where T : MonoBehaviour" works?

### ISSUE THREAD

**Issue by fitbie** [onewheelstudio/Adventures-in-C-Sharp#3]
Hi! I watched the generics video and had a question about how it works under the hood.

If you take the Pool<T> class from the tutorial and then in another class do something like 
```
    private GameManager gameManager;
    private PlayerController playerController;

    private void Example()
    {
        Pool<GameManager>.ReturnObject(gameManager);
        Pool<PlayerController>.ReturnObject(playerController);

        PlayerController secondPlayerController = Pool<PlayerController>.GetObject();
    }
```
And since we are dealing with Queue, we will get a GameManager instead of a PlayerController.
I understand the principle of using Generic for non-static classes, it speeds up the work with the code.
I understand the use of static classes with predefined type. 
But I don't understand the use of static generic pool classes, because if the class is static - it exists in one instance, which means that when you add different (because generic) types - chaos will start
Explain please, thanks!

**Comment by onewheelstudio** [comment]
Hey there.

Each instance of the object pool stores a different type. When you declare
a pool you must specify the type that is inside the pool. So something like
this

*Pool<GameObject> gameObjectPool;*
*Pool<Sprite> spritePool;*


This creates 2 instances of the pool class. This works even if the instance
is static.

Jeremy

**Comment by fitbie** [comment]
Yeah, so there is a problem: if i try to create instance of **static** class or using constructors
```
    private static Pool<Example> example1;
    or
    private Pool<Example> example2;
```
 i get error "Cannot declare static type variable ...."
And this makes sense for me, static class cannot be instantiated, so it would work perfect if the class wasn't static, or if it wasn't Generic.

I think I don't know something, can you help me figure it out?

**Comment by onewheelstudio** [comment]
You'll need to define the class first. So something like this.

public class ExamplePool : Pool<Example>
{

}

public class Example
{
}

Then you can do things like

public class SomeClass
{
     private static ExamplePool examplePool;
}

Hope that helps!

**Comment by fitbie** [comment]
Im totally agreed
But in your video there was public **static** class
So this was my problem, here is the link with timecode
https://youtu.be/6ED4Qo0Yi6o?t=658
And i was thinking thats all we need, without inheritance. Anyway, thank you!
![image](https://github.com/onewheelstudio/Adventures-in-C-Sharp/assets/119941436/95ccfc29-dd72-473d-9741-e8292fc32782)

**Comment by onewheelstudio** [comment]
Sorry my apologies, I saw the reference to "pool" and my brain assumed we
were talking about my object pooling video (which makes use of generics).

As for what's going on under the hood, to be honest, I'm not 100% sure what
the compiler and all are doing. My assumption is that when you do something
like Pool<SomeType>.SomeFunction() that the instance of the Pool<SomeType>
is being created. So when we do it multiple time with different types we
get multiple instances each with a different type stored in it.

As a simpler example of this working you can try the code below (I created
this to double check I hadn't made a major mistake in the video). Each
"pool" of value appears as a unique instance.

public class Playground : MonoBehaviour
{
    private void Awake()
    {
        Pool<int>.SomeFunction(12);
        Pool<float>.SomeFunction(0.5f);

        Debug.Log($"Object stored: {Pool<int>.someObject.ToString()}");
        Debug.Log($"Object stored: {Pool<float>.someObject.ToString()}");
    }
}

public static class Pool<T>
{
    public static T someObject;
    public static void SomeFunction(T someObject)
    {
        Pool<T>.someObject = someObject;
    }
}

**Comment by fitbie** [comment]
Im actually create a topic on UnityDiscussion, if youre intrested
https://discussions.unity.com/t/how-static-generic-classes-work-does-they-store-value-for-every-t/263550
